From 26280b92dfe6d66c72d8a811cc4930e21a6ffcf0 Mon Sep 17 00:00:00 2001
From: Ravindra Lokhande <rlokhande@nvidia.com>
Date: Thu, 9 Aug 2012 22:42:05 +0530
Subject: [PATCH] tinyalsa: return card number from its name substring

tinyalsa doesn't provide any way to get card number from its name.
Added function to return card number from card name substring. This
is needed as usb audio device can be plug-in and plug-out anytime
which changes card number.

Bug 1009921
Bug 1026047

Id-Google: I2967ede50392e22f2d6f94d0b785209e45cec94f
Reviewed-on: http://git-master/r/163879
Reviewed-by: Automatic_Commit_Validation_User
Tested-by: Nikesh Oswal <noswal@nvidia.com>
Reviewed-by: Ravindra Lokhande <rlokhande@nvidia.com>
Reviewed-by: Dan Willemsen <dwillemsen@nvidia.com>
Change-Id: Id72c42dc9ec2b8115416a6caedfe1b022001d1e9
Reviewed-on: http://git-master/r/329849
Reviewed-by: Riham Haidar <rhaidar@nvidia.com>
Tested-by: Riham Haidar <rhaidar@nvidia.com>
---
 include/tinyalsa/asoundlib.h |    3 +++
 mixer.c                      |   32 ++++++++++++++++++++++++++++++++
 2 files changed, 35 insertions(+)

diff --git a/include/tinyalsa/asoundlib.h b/include/tinyalsa/asoundlib.h
index 75e5cda..010d7bb 100644
--- a/include/tinyalsa/asoundlib.h
+++ b/include/tinyalsa/asoundlib.h
@@ -295,6 +295,9 @@ int mixer_ctl_set_enum_by_string(struct mixer_ctl *ctl, const char *string);
 int mixer_ctl_get_range_min(struct mixer_ctl *ctl);
 int mixer_ctl_get_range_max(struct mixer_ctl *ctl);
 
+/* Get sound card number using sub string */
+int snd_card_get_card_id_from_sub_string(const char *card_name);
+
 #if defined(__cplusplus)
 }  /* extern "C" */
 #endif
diff --git a/mixer.c b/mixer.c
index 4568cca..54d826e 100644
--- a/mixer.c
+++ b/mixer.c
@@ -495,3 +495,35 @@ int mixer_ctl_set_enum_by_string(struct mixer_ctl *ctl, const char *string)
     return -EINVAL;
 }
 
+
+int snd_card_get_card_id_from_sub_string(const char *card_name)
+{
+    int card;
+    int fd;
+    char fn[256];
+    struct snd_ctl_card_info info;
+
+    if (card_name == NULL)
+        return -EINVAL;
+
+    for (card = 0; card < 8; card++) {
+        snprintf(fn, sizeof(fn), "/dev/snd/controlC%u", card);
+        fd = open(fn, O_RDONLY);
+        if (fd < 0)
+            continue;
+        if (ioctl(fd, SNDRV_CTL_IOCTL_CARD_INFO, &info) < 0) {
+            close(fd);
+            continue;
+        }
+        close(fd);
+
+        /* search by card id */
+        if (strcasestr((const char *)info.id, card_name))
+            return info.card;
+
+        /* search by card name */
+        if (strcasestr((const char *)info.name, card_name))
+            return info.card;
+    }
+    return -EINVAL;
+}
-- 
1.7.9.5

